### DISCUSSION TITLE
Use something other than stdin for input?

### DISCUSSION THREAD

**Discussion by haliphax** [Textualize/textual#3493] [Q&A]
I'm trying to expose a Textual `App` within an SSH terminal server I'm building. As such, `stdin` and `stdout` in the Python environment belong to the _server_, not the _client_ connecting to the service; those are actually `IO` objects (asyncio `Queue`s under the hood, I believe).

I've managed to get the client's `stdout` taken care of by way of a custom `Driver` implementation and overridden `write` and `flush` methods, but I'm not seeing anything in there that would let me do the opposite for input (e.g. `read`). Methods such as rich's `Prompt.ask` provide a way to accomplish this, so it would be wonderful if the same was true of Textual.

Since this is a single-threaded, single-process asyncio server, monkey-patching `stdin` is not an option. That would not only break the server's `stdin`, but would also not work well when multiple clients are connected simultaneously with their own, individual sessions.

**Comment by willmcgugan** [comment]
A subprocess would be the typical solution. No need to write a driver when you can launch a process and manage its stdin and stdout.

**Reply by haliphax** [reply]
Yargh. That's what I was hoping to avoid. It may be easier to do with Textual than it was with Blessed, at least.

**Reply by haliphax** [reply]
Follow-up: The description of the `App.get_driver_class` method implies that a `Driver` is used for input and output, but I don't actually see any input management aside from `disable_input`. Are there plans for implementing any such methods?

**Reply by haliphax** [reply]
I managed to get it to work without a subprocess by making a subclass of `LinuxDriver` with an overridden `run_input_thread` method and some other bits and bobs. I haven't tested to see if it respects the terminal type of the connecting client rather than that of the server yet, but this is very promising! I hope to summarize my approach once it's been battle tested a bit longer and post it somewhere useful.

https://github.com/Textualize/textual/assets/745888/a5f135fd-7e41-4846-bc8a-49f5ffdf3d99

**Reply by dbieber** [reply]
Checking in if you intend to summarize and post still

**Reply by haliphax** [reply]
I haven't written up anything formal, but I can at the very least link you to the source code.

First off, here is a `FileWrapper` class I'm using to present a monkey-typed, file-like object to Textual: https://github.com/haliphax/xthulu/blob/c6af0ff3448621d504f0d4a17b4b1260ab77c7cd/xthulu/ssh/console/internal/file_wrapper.py

Next, here is an `SSHDriver` class which is derived from the `LinuxDriver` class in Textual (wrapping the `SSHWriter` object provided by `asyncssh` for my particular use case) where I am handling the input and output: https://github.com/haliphax/xthulu/blob/c6af0ff3448621d504f0d4a17b4b1260ab77c7cd/xthulu/ssh/console/internal/driver.py

Finally, here is my custom `Console` subclass that incorporates the mechanics from above: https://github.com/haliphax/xthulu/blob/c6af0ff3448621d504f0d4a17b4b1260ab77c7cd/xthulu/ssh/console/__init__.py
